src/v1/app.files.ts

```diff
@@ -256,6 +257,31 @@ export function initFileTree() {
         return false;
     });
 
+    uiFiles.dropdownSubmenus.mouseenter(e => {
+        var el = $(e.currentTarget);
+        if (!el.hasClass("disabled")) {
+            var menu = el.find("> .dropdown-menu");
+            menu.css({ display: "block", visibility: "hidden", "z-index": -1 });
+            var itemPos = el.offset();
+            var menuW = menu.outerWidth();
+            var menuH = menu.outerHeight();
+            var x = itemPos.left + el.width() + menuW <= $(window).width() ? itemPos.left + el.width() : itemPos.left - menuW;
+            var y = itemPos.top + menuH <= $(window).height()
+                ? itemPos.top
+                : itemPos.top >= menuH
+                    ? itemPos.top + el.height() - menu.height()
+                    : $(window).height() - menuH;
+            x -= itemPos.left;
+            y -= itemPos.top;
```

Comment:
Couldn't `itemPos.left` and `itemPos.top` just be left out of the calculations for `x` and `y`?

Reply:
From the mathematical POV it probably could, but I think this expresses the intent better. Variables `x` and `y` store absolute positions from the top left corner of the browser window (in pixels), but they can't be applied directly as values of the CSS properties `left` and `top` as the dropdown submenu is positioned relatively to its parent menu item (because [every `.dropdown-submenu` has `position: relative`](https://github.com/kaitai-io/kaitai_struct_webide/blob/a9b5e580c191f67abe4115815a6db7646fe0af01/css/contextmenu.css#L2)), so I'm subtracting the offset of the from the absolute position to get the relative one. See the bounding box of the menu item _Generate parser_ with class `dropdown-submenu`:
![dropdown-submenu](https://user-images.githubusercontent.com/47499687/76689380-978b2580-6635-11ea-930b-4dafc550e132.gif)

I tried to remove the `position: relative` from the rule `.dropdown-submenu` to be able to position the dropdown menu absolute to the window, but it didn't work well, because the whole context menu has indeed `position: absolute`, so all its descendants are positioned relative to its offset and IMHO there isn't a sane way to "escape" from that box. It's easier to calculate the position of the dropdown menu relative to the parent item.

However, I still think it's easier to think in absolute positions relative to the browser window instead of realizing which element are the positions relative to. As I said, I think this approach expresses the intent better, and I don't think the expressions would simplify when leaving out the `itemPos` from the calculations (we couldn't do it completely anyway and hardly anyone would know what those expressions originated from).

Reply:
Makes sense, thanks for the explanation.